libs/wingc/src/type_check/fqn.rs

```diff
@@ -0,0 +1,152 @@
+use std::fmt::{Display, Formatter};
+
+use crate::{CONSTRUCT_BASE, WINGSDK_ASSEMBLY_NAME, WINGSDK_RESOURCE};
+
+/// Represents a fully-qualified name (FQN) of a type in a JSII library.
+/// For example, `@aws-cdk/aws-ec2.Vpc` is a FQN.
+/// The FQN uniquely identifies a type.
+///
+/// The FQN typically looks like:
+/// `assembly_name.namespace1.namespace2...namespaceN.type_name`
+/// where there may be zero or more namespaces.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct FQN<'a>(&'a str);
+
+impl<'a> From<&'a str> for FQN<'a> {
+	fn from(fqn: &'a str) -> Self {
+		if fqn.split('.').count() < 2 {
+			panic!("Invalid FQN: {}", fqn);
+		}
+		FQN(fqn)
+	}
+}
+
+impl<'a> FQN<'a> {
+	/// Returns the FQN as a string.
+	pub fn as_str(&self) -> &str {
+		self.0
+	}
+
+	/// Returns the FQN as a string without the type name.
+	pub fn as_str_without_type_name(&self) -> &str {
+		let index = self.0.rfind('.').unwrap();
+		&self.0[..index]
+	}
+
+	/// Returns the "assembly" part of the FQN. This is the name of the
+	/// npm library or wing library the type is defined in.
+	#[allow(dead_code)]
+	pub fn assembly(&self) -> &str {
+		self.0.split('.').next().unwrap()
+	}
+
+	/// Returns the "type name" part of the FQN. This is the name of the
+	/// type itself.
+	pub fn type_name(&self) -> &str {
+		self.0.split('.').last().unwrap()
+	}
+
+	/// Returns an iterator over the namespaces of the FQN. The namespaces
+	/// are the parts of the FQN between the assembly name and the type name
+	/// which are used to group types together.
+	pub fn namespaces(&self) -> impl Iterator<Item = &str> {
+		let mut parts = self.0.split('.');
+		parts.next();
+		parts.next_back().unwrap();
+		parts
+	}
+
+	/// Returns true if the FQN represents a "construct base class".
+	///
+	/// TODO: this is a temporary hack until we support interfaces.
+	pub fn is_construct_base(&self) -> bool {
```

Comment:
Feels weird that this is under `FQN`... Makes more sense as a free-floating function to me.

Reply:
👍